internal/controller/bmc/controller.go

```diff
@@ -0,0 +1,274 @@
+			}),
+		).
+		Watches(
+			&corev1.Secret{},
+			handler.EnqueueRequestsFromMapFunc(func(ctx context.Context, obj client.Object) []ctrl.Request {
+				var list bmcv1beta1.VirtualMachineBMCList
+				if err := r.List(ctx, &list, client.MatchingFields{"spec.authSecretRef.name": obj.GetName()}); err != nil {
+					return nil
+				}
+				var reqs []ctrl.Request
+				for _, item := range list.Items {
+					reqs = append(reqs, ctrl.Request{NamespacedName: types.NamespacedName{
+						Name: item.Name, Namespace: item.Namespace,
+					}})
+				}
+				return reqs
+			}),
+		).
+		Complete(r)
+}
+
+func (r *VirtualMachineBMCReconciler) UpdateStatusIfNeeded(ctx context.Context, virtBMC *bmcv1beta1.VirtualMachineBMC, clusterIP string, condition metav1.Condition) error {
+	updated := false
+
+	if clusterIP != "" && virtBMC.Status.ClusterIP != clusterIP {
+		virtBMC.Status.ClusterIP = clusterIP
+		updated = true
+	}
+
+	found := false
+	for i, cond := range virtBMC.Status.Conditions {
+		if cond.Type == condition.Type {
+			found = true
+			if cond.Status != condition.Status || cond.Message != condition.Message {
+				virtBMC.Status.Conditions[i] = condition
+				updated = true
+			}
+			break
+		}
+	}
+	if !found {
+		virtBMC.Status.Conditions = append(virtBMC.Status.Conditions, condition)
+		updated = true
+	}
+
+	if updated {
+		r.Log.Info("Updating VirtualMachineBMC status")
+		if err := r.Status().Update(ctx, virtBMC); err != nil {
+			r.Log.Error(err, "Failed to update VirtualMachineBMC status")
+			return err
+		}
+	}
+
+	return nil
+}
+
+func (r *VirtualMachineBMCReconciler) ValidateRefs(ctx context.Context, virtBMC *bmcv1beta1.VirtualMachineBMC) (ctrl.Result, error) {
+	if virtBMC == nil {
+		return ctrl.Result{}, errors.New("VirtualMachineBMC is nil")
+	}
+
+	if virtBMC.Spec.VirtualMachineRef == nil || virtBMC.Spec.VirtualMachineRef.Name == "" {
+		r.Log.Error(nil, "VirtualMachineRef is required")
+		return ctrl.Result{}, nil
+	}
+
+	var vm kubevirtv1.VirtualMachine
+	if err := r.Get(ctx, types.NamespacedName{
+		Name:      virtBMC.Spec.VirtualMachineRef.Name,
+		Namespace: virtBMC.Namespace,
+	}, &vm); err != nil {
+		if apierrors.IsNotFound(err) {
+			return ctrl.Result{}, nil
+		}
+		return ctrl.Result{}, err
+	}
+
+	if virtBMC.Spec.AuthSecretRef == nil || virtBMC.Spec.AuthSecretRef.Name == "" {
+		r.Log.Error(nil, "AuthSecretRef is required")
+		return ctrl.Result{}, nil
+	}
+
+	var secret corev1.Secret
+	if err := r.Get(ctx, types.NamespacedName{
+		Name:      virtBMC.Spec.AuthSecretRef.Name,
+		Namespace: virtBMC.Namespace,
+	}, &secret); err != nil {
+		if apierrors.IsNotFound(err) {
+			return ctrl.Result{}, nil
+		}
+		return ctrl.Result{}, err
+	}
+
+	return ctrl.Result{}, nil
+}
+
+func (r *VirtualMachineBMCReconciler) HandleFinalizer(ctx context.Context, virtBMC *bmcv1beta1.VirtualMachineBMC) (ctrl.Result, error) {
+	if virtBMC.ObjectMeta.DeletionTimestamp.IsZero() {
+		if !controllerutil.ContainsFinalizer(virtBMC, BMCFinalizer) {
```

Comment:
Why use a finalizer again? Can't you set owner references on the Deployment and Service that would result in them getting deleted automatically when the BMC CR is deleted?

Reply:
Why do you add a finalizer if it doesn't do anything?

Reply:
I don't think adding a finalizer to the VirtualMachineBMC objects explicitly is adding any value. Like @0xFelix said, adding OwnerReferences to the Deployment and Service object should be enough.